### PR TITLE
Feature: Kahan-Babuska-Neumaier compensated sum

### PR DESCRIPTION
**Pull request by liss-h**
Implement compensated sum for SPARQL `SUM()` aggregate.

Benefits:
- less error accumulated throughout the sum
- sum results are more often inlineable

### CODE REVIEW

File: src/rdf4cpp/CompensatedSum.hpp [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+ * }
+ * Literal const result = sum.value();
+ * @endcode
+ */
+struct CompensatedSum {
+private:
+    storage::DynNodeStoragePtr node_storage_;
+
+    std::optional<DeferredValue> sum_;  // running total (initially zero)
+    DeferredValue comp_;                // low-order bits lost by the additions so far, only valid if !empty_
```
Comment (nkaralis) [inline src/rdf4cpp/CompensatedSum.hpp]: what is `empty_`?

File: src/rdf4cpp/datatypes/registry/DatatypeRegistry.hpp [hunk 2]
```diff
@@ -569,12 +577,15 @@ inline void DatatypeRegistry::add() noexcept {
     using conversion_table_t = decltype(make_conversion_table_for<LiteralDatatype_t>());
 
     auto const num_ops = []() -> std::optional<NumericOps> {
+        // note: not std::floating_point, that would miss owl:real whose cpp_type is a class type
```
Comment (nkaralis) [inline src/rdf4cpp/datatypes/registry/DatatypeRegistry.hpp]: is this referring to the usage of numeric_limits?

File: src/rdf4cpp/CompensatedSum.cpp [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+void CompensatedSum::add(Literal const &lit) {
+    this->add(make_deferred_from_literal(lit));
+}
+
+void CompensatedSum::add(DeferredValue const &value) {
+    if (!sum_.has_value()) {
+        // the first value seeds the sum; adding it to a "0"^^xsd:integer instead would poison
+        // owl:rational and owl:real, which have no common numeric type with xsd:integer
+        sum_ = value;
+        comp_ = numeric_sub_deferred(value, value, node_storage_);  // zero, in the datatype of value
```
Comment (nkaralis) [inline src/rdf4cpp/CompensatedSum.cpp]: why is this needed?

File: src/rdf4cpp/CompensatedSum.cpp [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+        compensating_ = !this->is_exact(value.second);
+        return;
+    }
+
+    if (!compensating_ && this->is_exact(value.second)) {
```
Comment (nkaralis) [inline src/rdf4cpp/CompensatedSum.cpp]: style-related: `compensating_` without `this-` vs `this->exact` (feel free to skip it)

File: tests/nodes/tests_CompensatedSum.cpp [hunk 5]
```diff
@@ -0,0 +1,228 @@
...
+        // Kahan loses the compensation when |value| >> |sum|; Neumaier would return 1.0 here
+        std::vector<Literal> const lits{Literal::make_typed_from_value<datatypes::xsd::Double>(1e16),
+                                        Literal::make_typed_from_value<datatypes::xsd::Double>(1.0),
+                                        Literal::make_typed_from_value<datatypes::xsd::Double>(-1e16)};
+
+        CHECK_EQ(compensated_sum(lits).value<datatypes::xsd::Double>(), 0.0);
+    }
+
+    SUBCASE("an infinite value") {
+        // the correction term becomes inf - inf = NaN and poisons the total, where operator+ keeps
```
Comment (nkaralis) [inline tests/nodes/tests_CompensatedSum.cpp]: i think this should be addressed in this PR

**Review by nkaralis — COMMENTED** [review]
Thank you for the PR. I have some comment.

File: src/rdf4cpp/CompensatedSum.hpp (hunk 1 above)
Comment (liss-h) [reply src/rdf4cpp/CompensatedSum.hpp]: leftover comment. Will fix

File: src/rdf4cpp/datatypes/registry/DatatypeRegistry.hpp (hunk 2 above)
Comment (liss-h) [reply src/rdf4cpp/datatypes/registry/DatatypeRegistry.hpp]: removed comment

File: src/rdf4cpp/DeferredLiteral.cpp [hunk 6]
```diff
@@ -0,0 +1,105 @@
...
+            lhs, rhs, node_storage);
+}
+
+// Fast2Sum (Dekker): for |a| >= |b|, (a - (a + b)) + b is exactly the rounding error of a + b.
+// Adding those errors back up is what makes a running total compensated (Kahan-Babuska-Neumaier
+// summation, see CompensatedSum). The exactness only holds within one floating point format, which
+// is why the transformation runs on the equalized values rather than on the two DeferredValues.
+// See https://en.wikipedia.org/wiki/2Sum
+NumericAddWithLossResult numeric_add_with_loss_deferred(DeferredLiteral const &lhs,
+                                                                       DeferredLiteral const &rhs,
```
Comment (nkaralis) [inline src/rdf4cpp/DeferredLiteral.cpp]: fmt is off here

File: src/rdf4cpp/DeferredLiteral.hpp [hunk 7]
```diff
@@ -0,0 +1,186 @@
...
+[[nodiscard]] DeferredLiteral numeric_mul_deferred(DeferredLiteral const &lhs,
+                                                   DeferredLiteral const &rhs,
+                                                   storage::DynNodeStoragePtr node_storage = storage::default_node_storage);
+[[nodiscard]] DeferredLiteral numeric_div_deferred(DeferredLiteral const &lhs,
+                                                   DeferredLiteral const &rhs,
+                                                   storage::DynNodeStoragePtr node_storage = storage::default_node_storage);
+
+struct NumericAddWithLossResult {
+    DeferredLiteral result;
+    DeferredLiteral loss;
```
Comment (nkaralis) [inline src/rdf4cpp/DeferredLiteral.hpp:170]: does this need to expose `null()`?

File: src/rdf4cpp/DeferredLiteral.cpp [hunk 8]
```diff
@@ -0,0 +1,105 @@
...
+        // neither add nor sub changes the datatype, so these same ops apply to the sum
+        IRI const datatype{sum.result_type_id, node_storage};
+
+        auto loss = [&]() -> DeferredLiteral {
+            if (num_ops.is_inf_fptr(*sum.result_value)) {
+                return {};  // an infinite sum has nothing to report but inf - inf = NaN
+            }
+
+            // Fast2Sum wants the operand of larger magnitude first. Operands that do not order
+            // (NaN, or a datatype that cannot compare) may go either way: the loss is then merely
```
Comment (nkaralis) [inline src/rdf4cpp/DeferredLiteral.cpp:82]: is the case "operands that do not order" tested?

File: src/rdf4cpp/CompensatedSum.cpp [hunk 9]
```diff
@@ -0,0 +1,74 @@
...
+    if (!compensating_ && is_exact(value.datatype)) {
+        // arbitrary precision, so there is no rounding error to carry
+        sum_ = numeric_add_deferred(*sum_, value, node_storage_);
+        return;
+    }
+
+    compensating_ = true;
+
+    // Kahan-Babuska-Neumaier: collect the loss of each addition on the side and add it back in
+    // value().
```
Comment (nkaralis) [inline src/rdf4cpp/CompensatedSum.cpp]: fmt: the comment does not a new line

File: src/rdf4cpp/CompensatedSum.cpp [hunk 10]
```diff
@@ -0,0 +1,74 @@
...
+    auto [new_sum, loss] = numeric_add_with_loss_deferred(*sum_, value, node_storage_);
+
+    // a null loss is one there is nothing to account for: the total went infinite, or it is
+    // poisoned and t carries that on. comp_ itself is the null-value until the first loss seeds it
+    if (!loss.null()) {
+        comp_ = comp_.null() ? loss : numeric_add_deferred(comp_, loss, node_storage_);
+    }
+
+    sum_ = std::move(new_sum);
```
Comment (nkaralis) [inline src/rdf4cpp/CompensatedSum.cpp:72]: what happens here if `loss.null()` is true?

File: src/rdf4cpp/CompensatedSum.cpp [hunk 11]
```diff
@@ -0,0 +1,74 @@
...
+    }
+
+    compensating_ = true;
+
+    // Kahan-Babuska-Neumaier: collect the loss of each addition on the side and add it back in
+    // value().
+    auto [new_sum, loss] = numeric_add_with_loss_deferred(*sum_, value, node_storage_);
+
+    // a null loss is one there is nothing to account for: the total went infinite, or it is
+    // poisoned and t carries that on. comp_ itself is the null-value until the first loss seeds it
```
Comment (nkaralis) [inline src/rdf4cpp/CompensatedSum.cpp]: typo; t -> it

**Review by nkaralis — COMMENTED** [review]
Thank you for the updates. I have some questions and minor comments

File: src/rdf4cpp/CompensatedSum.cpp [hunk 12]
```diff
@@ -0,0 +1,74 @@
...
+
+void CompensatedSum::add(Literal const &lit) {
+    this->add(make_deferred_from_literal(lit));
+}
+
+void CompensatedSum::add(DeferredLiteral const &value) {
+    if (!sum_.has_value()) {
+        // the first value seeds the sum; adding it to a "0"^^xsd:integer instead would poison
+        // owl:rational and owl:real, which have no common numeric type with xsd:integer
+        sum_ = value;
```
Comment (nkaralis) [inline src/rdf4cpp/CompensatedSum.cpp:50]: Here, a non numeric literal should make the `sum_` null.
```
sum.add(Literal::make_simple("abc"));
sum.value();   // "abc"^^xsd:string
```
We should add test cases that check what happens if a non-numeric literal is added

File: src/rdf4cpp/DeferredLiteral.hpp [hunk 13]
```diff
@@ -0,0 +1,186 @@
...
+#include <rdf4cpp/Literal.hpp>
+#include <rdf4cpp/IRI.hpp>
+
+#include <any>
+
+namespace rdf4cpp {
+
+/**
+ * @brief The value of a Literal that is not (yet) placed into a node storage, together with the IRI of its datatype.
+ * A DeferredValue with a null datatype IRI is the null-value; it is what the numeric_*_deferred functions
```
Comment (nkaralis) [inline src/rdf4cpp/DeferredLiteral.hpp]: DiferredValue -> DifferedLiteral

File: tests/nodes/tests_CompensatedSum.cpp [hunk 14]
```diff
@@ -0,0 +1,267 @@
...
+TEST_CASE("errors poison the sum") {
+    SUBCASE("non-numeric value") {
+        CompensatedSum sum;
+        sum.add(Literal::make_typed_from_value<datatypes::xsd::Double>(1.0));
+        sum.add(Literal::make_simple("not a number"));
+
+        CHECK(sum.value().null());
+    }
+
+    SUBCASE("null literal") {
```
Comment (nkaralis) [inline tests/nodes/tests_CompensatedSum.cpp:173]: please add a subcase with `DeferredLiteral{}`

File: tests/nodes/tests_CompensatedSum.cpp [hunk 15]
```diff
@@ -0,0 +1,267 @@
...
+        std::vector<Literal> const lits{Literal::make_typed_from_value<datatypes::xsd::Decimal>(BigDecimal<>{"0.5"}),
+                                        Literal::make_typed_from_value<datatypes::xsd::Float>(0.25f)};
+
+        CHECK_EQ(compensated_sum(lits).datatype(), naive_sum(lits).datatype());
+        CHECK_EQ(compensated_sum(lits).datatype(), IRI{datatypes::xsd::Float::identifier});
+    }
+}
+
+TEST_CASE("errors poison the sum") {
+    SUBCASE("non-numeric value") {
```
Comment (nkaralis) [inline tests/nodes/tests_CompensatedSum.cpp:149]: here, we need a test case that starts with a non-numeric literal

File: tests/nodes/tests_CompensatedSum.cpp [hunk 16]
```diff
@@ -0,0 +1,267 @@
...
+}
+
+template<datatypes::LiteralDatatype T>
+std::vector<Literal> repeat(typename T::cpp_type const &value, size_t n) {
+    return std::vector<Literal>(n, Literal::make_typed_from_value<T>(value));
+}
+
+}  // namespace
+
+TEST_CASE("empty sum is zero") {
```
Comment (nkaralis) [inline tests/nodes/tests_CompensatedSum.cpp:43]: suggestions for tests cases if you think are needed:
- check that `comp_` survives promotion
- dedicated checks for the results of `numeric_add_with_loss_deferred`

File: src/rdf4cpp/DeferredLiteral.cpp (hunk 8 above)
Comment (liss-h) [reply src/rdf4cpp/DeferredLiteral.cpp:82]: No, we don't have any types that do this

File: src/rdf4cpp/DeferredLiteral.hpp (hunk 7 above)
Comment (liss-h) [reply src/rdf4cpp/DeferredLiteral.hpp:170]: No, result can be non-null while loss is null

File: src/rdf4cpp/CompensatedSum.cpp (hunk 10 above)
Comment (liss-h) [reply src/rdf4cpp/CompensatedSum.cpp:72]: no loss could be calculated for this step and therefore is treated as 0 loss

File: src/rdf4cpp/CompensatedSum.cpp [hunk 17]
```diff
@@ -25,11 +25,31 @@ bool CompensatedSum::is_exact(IRI const &datatype) {
...
+void CompensatedSum::add(Literal const &lit, uint64_t multiplicity) {
+    this->add(make_deferred_from_literal(lit), multiplicity);
 }
 
-void CompensatedSum::add(DeferredLiteral const &value) {
+// value * multiplicity as the sum of value * 2^i over the set bits of multiplicity. Doubling is exact
+// in binary floating point (bar overflow, where the product overflows too), so every term is exact and
+// only the compensated additions round
+void CompensatedSum::add(DeferredLiteral const &value, uint64_t multiplicity) {
+    if (multiplicity == 1) [[likely]] {
```
Comment (nkaralis) [inline src/rdf4cpp/CompensatedSum.cpp]: should we remove `[[likely]]`?

File: src/rdf4cpp/CompensatedSum.cpp [hunk 18]
```diff
@@ -0,0 +1,113 @@
...
+
+void CompensatedSum::add(Literal const &lit, uint64_t multiplicity) {
+    this->add(make_deferred_from_literal(lit), multiplicity);
+}
+
+// value * multiplicity as the sum of value * 2^i over the set bits of multiplicity. Doubling is exact
+// in binary floating point (bar overflow, where the product overflows too), so every term is exact and
+// only the compensated additions round
+void CompensatedSum::add(DeferredLiteral const &value, uint64_t multiplicity) {
+    if (multiplicity == 1) [[likely]] {
```
Comment (nkaralis) [inline src/rdf4cpp/CompensatedSum.cpp]: we should probably remove [[likely]]

File: tests/nodes/tests_CompensatedSum.cpp (hunk 16 above)
Comment (liss-h) [reply tests/nodes/tests_CompensatedSum.cpp:43]: I added a test that starts out with float and ends up on double.
I don't think we need additional tests for numeric_add_with_loss_deferred itself

File: src/rdf4cpp/CompensatedSum.cpp (hunk 18 above)
Comment (liss-h) [reply src/rdf4cpp/CompensatedSum.cpp]: really? Ok, I was under the impression that multiplicity 1 is much more likely. Will remove

File: tests/nodes/tests_CompensatedSum.cpp [hunk 19]
```diff
@@ -410,3 +410,13 @@ TEST_CASE("near max") {
 
     CHECK_EQ(s.value(), manual_sum);
 }
+
+TEST_CASE("poisoned") {
+    CompensatedSum s;
+    s.add(1_xsd_integer);
+    CHECK_FALSE(s.poisoned());
+
+    s.add("hello"_xsd_string);
```
Comment (nkaralis) [inline tests/nodes/tests_CompensatedSum.cpp:428]: can you also add a test case with only the string?

**Review by nkaralis — APPROVED** [review]
Thank you for the latest changes.
